// _redux/reducers/game.settings.reducer.js
import {gameSettingsConstants} from "_redux/constants/game.constants";

const initialState = {
    modes: [],
    status: {
        mode: false,
        isActive: false,
        player: ''
    },
    pending: false,
    fetched: false,
};

export function gameSettings (state = initialState, {payload, type}) {
    switch (type) {
        case gameSettingsConstants.GET.REQUEST:
            return {
                ...state,
                pending: true,
            };
        case gameSettingsConstants.GET.SUCCESS:
            return {
                ...state,
                modes: payload,
                fetched: true,
                pending: true,
            };
        case gameSettingsConstants.GET.ERROR:
            return {
                ...state,
                pending: false,
            };
        case gameSettingsConstants.SET:
            return {
                ...state,
                status: {
                    ...payload,
                    isActive: true
                }
            };
        default:
            return state;
    }
}
